File: week-02/src/main/java/org/distance/GuessValue.java
package org.distance;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;

public class GuessValue {
    private final Map<String, Integer> valueIndexMap = new HashMap<>();
    private final int alphabetSize;
    private final String target;

    public enum GuessResult {
        ICE("Ice! You're very far from the target."),
        COLD("Cold! You're a bit far off."),
        WARM("Warm! You're getting there."),
        HOT("Hot! You're very close."),
        MATCH("Congratulations! You guessed the correct value."),
        NONE("No valid guess made.");

        private final String message;

        GuessResult(String message) {
            this.message = message;
        }

        public String getMessage() {
            return message;
        }
    }

    public GuessValue(String[] guessValues) throws IOException {
        this.alphabetSize = guessValues.length;
        this.target = ConfigLoader.getTargetValue();

        // To avoid repeatedly converting the array to list
        for (int i = 0; i < guessValues.length; i++) {
            this.valueIndexMap.put(guessValues[i].toLowerCase(), i);
        }

    }

    public GuessResult findHowClose(String guess) throws IOException {
        if (target == null || target.isEmpty()) return GuessResult.NONE;
        if (guess == null || guess.isEmpty()) return GuessResult.NONE;

        int indexOfGuess = valueIndexMap.get(guess);
        int indexOfTarget = valueIndexMap.get(target.toLowerCase());

        int distance = Math.abs(indexOfGuess - indexOfTarget);
        double fraction = (double) (distance) / (alphabetSize);

        if (fraction == 0) return GuessResult.MATCH;
        if (fraction < 0.25) return GuessResult.HOT;
        if (fraction < 0.5) return GuessResult.WARM;
        if (fraction < 0.75) return GuessResult.COLD;

        return GuessResult.ICE;

    }

}

File: guess-algorithm/typescript/guessValue.ts

const target = "k";
const map = new Map<string, number>();
const GuessResult = {
  ICE: "Ice! You're very far from the target.",
  COLD: "Cold! You're a bit far off.",
  WARM: "Warm! You're getting there.",
  HOT: "Hot! You're very close.",
  MATCH: "Congratulations! You guessed the correct value.",
  NONE: "No valid guess made.",
} as const;

type GuessResult = keyof typeof GuessResult;

function guessValue(guessValues: string[], guess: string): GuessResult {
    const size = guessValues.length;
    if(!target || size === 0 || !guess) {
        return "NONE";
    }
    for (let i = 0; i < guessValues.length; i++) {
        const value = guessValues[i];
        if (!map.has(value)) {
            map.set(value, i);
        }
    }

    if (!map.has(guess)) {
        return "NONE";
    }
    
    const distance = (map.get(target) || 0) - (map.get(guess) || 0);
    const fraction = Math.abs(distance) / size;

    if (fraction === 0) return "MATCH";
    if (fraction < 0.25) return "HOT";
    if (fraction < 0.5) return "WARM";
    if (fraction < 0.75) return "COLD";

    return "ICE";

}


const guessValues = ["a", "b", "c", "d", "e", "f", "g", "h", "i", "j", "k", "l", "m", "n", "o", "p", "q", "s", "t", "u", "v", "w", "x", "y", "z"];
const result: GuessResult = guessValue(guessValues, "g");
console.log(GuessResult[result])
